test/Chainweb/Test/Cut.hs

```diff
@@ -142,92 +205,64 @@ testMine'
-offsetBlockTime offset cut cid = add offset t
-  where
-    BlockCreationTime t = cut ^?! ixg cid . blockCreationTime
-
-arbitraryBlockTimeOffset
-    :: TimeSpan Micros
-    -> TimeSpan Micros
-    -> T.Gen GenBlockTime
-arbitraryBlockTimeOffset lower upper = do
-    t <- genEnum (lower, upper)
-    return $ offsetBlockTime t
+    try (createNewCut (chainLookupM wdb) n (t c (_chainId i)) payloadHash i c) >>= \case
+        Right p@(T2 h _) -> Right p <$ insertWebBlockHeaderDb wdb h
+        e -> return e
 
 testMineWithPayloadHash
-    :: forall cid
+    :: forall cid hdb
     . HasChainId cid
-    => Nonce
+    => ChainValueCasLookup hdb BlockHeader
+    => hdb
+    -> Nonce
     -> Time Micros
     -> BlockPayloadHash
     -> cid
     -> Cut
     -> IO (Either MineFailure (T2 BlockHeader Cut))
-testMineWithPayloadHash n t payloadHash i c =
-    forM (createNewCut n t payloadHash i c) return
+testMineWithPayloadHash db n t ph cid c = try
+    $ createNewCut (chainLookupM db) n t ph cid c
 
 -- | Create a new block. Only produces a new cut but doesn't insert it into the
 -- chain database.
 --
+-- The creation time isn't checked.
+--
 createNewCut
     :: HasCallStack
+    => MonadCatch m
     => HasChainId cid
-    => Nonce
+    => (ChainValue BlockHash -> m BlockHeader)
+    -> Nonce
     -> Time Micros
     -> BlockPayloadHash
     -> cid
     -> Cut
-    -> Either MineFailure (T2 BlockHeader Cut)
-createNewCut n t pay i c = do
-    h <- note BadAdjacents $ newHeader . BlockHashRecord <$> newAdjHashes
-    unless (checkTarget (_blockTarget h) $ _blockPow h) $ Left BadNonce
-    c' <- first (\e -> error $ "Chainweb.Cut.createNewCut: " <> sshow e)
-        $ monotonicCutExtension c h
+    -> m (T2 BlockHeader Cut)
+createNewCut hdb n t pay i c = do
+    extension <- fromMaybeM BadAdjacents $ getCutExtension c i
+    work <- newWorkHeaderPure hdb (BlockCreationTime t) extension pay
+    (h, mc') <- extendCut c pay (solveWork work n t)
+        `catch` \(InvalidSolvedHeader _ msg) -> throwM $ InvalidHeader msg
+    c' <- fromMaybeM BadAdjacents mc'
     return $ T2 h c'
-  where
-    cid = _chainId i
-
-    -- | The parent block to mine on.
-    --
-    p :: BlockHeader
-    p = c ^?! ixg cid
-
-    newHeader :: BlockHashRecord -> BlockHeader
-    newHeader as = newBlockHeader as pay n (BlockCreationTime t) $ ParentHeader p
-
-    -- | Try to get all adjacent hashes dependencies.
-    --
-    newAdjHashes :: Maybe (HM.HashMap ChainId BlockHash)
-    newAdjHashes = iforM (_getBlockHashRecord $ _blockAdjacentHashes p) $ \xcid _ ->
-        c ^?! ixg xcid . to (tryAdj (_blockHeight p))
-
-    tryAdj :: BlockHeight -> BlockHeader -> Maybe BlockHash
-    tryAdj h b
-        | _blockHeight b == h = Just $! _blockHash b
-        | _blockHeight b == h + 1 = Just $! _blockParent b
-        | otherwise = Nothing
 
 -- | Create a new cut where the new block has a creation time of one second
 -- after its parent.
 --
-createNewCutWithoutTime
```

Comment:
This is only used once, here. Maybe we should ditch this alternate and inline the call to `createNewCut`.